crates/node-bindings/src/file_system/file_system_napi.rs

```diff
@@ -1,94 +1,52 @@
 use std::path::Path;
-use std::path::PathBuf;
 
-use napi::bindgen_prelude::FromNapiValue;
-use napi::threadsafe_function::ThreadsafeFunctionCallMode;
-use napi::Env;
-use napi::JsFunction;
 use napi::JsObject;
 use parcel::file_system::FileSystem;
-use serde::de::DeserializeOwned;
-use serde::Serialize;
+
+use crate::helpers::js_callable::map_params_serde;
+use crate::helpers::js_callable::map_return_serde;
+use crate::helpers::js_callable::JsCallable;
 
 // TODO error handling
 
 pub struct FileSystemNapi {
-  read_file_fn: Box<dyn Fn((PathBuf, String)) -> String + Send + Sync>,
-  is_file_fn: Box<dyn Fn(PathBuf) -> bool + Send + Sync>,
-  is_dir_fn: Box<dyn Fn(PathBuf) -> bool + Send + Sync>,
+  read_file_fn: JsCallable,
```

Comment:
Can you move the result/param generics back to the variable itself;

Reply:
I tried this initially but it had a couple problems:

1. JsFunctions are overloadable, so it's better for the type signature to be defined at the call site
2. Functions that take multiple properties do so in the form of `Vec<JsUnknown>`. 

This can be problematic for a generic as I need to constrain the type to be iterable but also include tuples.

```rust
// T = impl Serialize

JsCallable<usize>
// This translates to: 
//    function(arg0: number): void

JsCallable<(usize, String)>
// This translates to: 
//    function(arg0: [number, string]): void
```

```rust
// T = Iterator<Item = impl Serialize>

JsCallable<Vec<usize>>
// This translates to: 
//    function(...args: number[]): void

JsCallable<(usize, String)> // This is not valid
```
I couldn't figure this out, so I just left the typing at the function call

Reply:
I don't think
```
JsCallable<(usize, String)>
```

Translates to `[number, string]` ; we are already keeping the generic at the construction site on the current `create_js_thread_safe_method` implementation

Reply:
That's okay, wanted to explain my reasoning. Can I mark this comment as resolved?

Reply:
I'm resolving even though I think it'd be slightly better to move the generic here:

* 2 less allocations per call
* actually less code